src/test/java/htsjdk/variant/vcf/VCFHeaderLineTranslatorUnitTest.java

```diff
@@ -14,49 +20,143 @@ public void testParseVCF4HeaderLine() {
+                                                                                  "<ID=ANNOTATION,Description=\"ANNOTATION \\n with a newline in it\\\">",
+                                                                                  null);
             Assert.fail("Should have thrown a TribbleException for having an unclosed quote in the description line");
-        } catch (TribbleException.InvalidHeader e) {
         }
+        catch (TribbleException.InvalidHeader e) {
+        }
+
+    }
+
+    @DataProvider(name = "testData")
+    private Object[][] getTestData() {
+        String line = "<ID=X,Description=\"Y\">";
+        return new Object[][]{
+                {line, "ID,Description", "", ""},
+                {line, "ID", "Description", ""},
+                {line, "ID", "Description,Optional2", ""},
+                {line, "", "Description,ID", ""},
+                {line, "Description,ID", "", "Wrong order of parameters"},
+                {line, "Description", "ID", "Wrong order of parameters"},
+                {line, "ID,Desc", "", "Parameter name mismatch"},
+                {line, "ID", "Desc", "Optional parameter name mismatch"},
+                {line, "ID,Description,Extra", "", ""},
+                {"<>", "", "", ""},
+                {"<>", "", "ID,Description", ""},
+                {"<>", "ID", "", ""},
+                {"<ID=X,Description=\"Y\",Extra=E>", "ID", "Description", "Too many parameters"},
+                {"<ID=X,Description=<Y>>", "ID,Description", "", ""}
+        };
+    }
+
+    @Test(dataProvider = "testData")
+    public void testParseVCF4HeaderLineWithTags(final String line,
+                                                final String expected,
+                                                final String optional,
+                                                final String error) {
+        try {
+            final List<String> expectedTagOrder = expected.isEmpty() ?
+                    Collections.emptyList() :
+                    Arrays.asList(expected.split(","));
+            final Set<String> optionalTags = optional.isEmpty() ?
+                    Collections.emptySet() :
+                    Sets.newHashSet(optional.split(","));
+            
+            if (optionalTags.isEmpty()) {
+                VCFHeaderLineTranslator.parseLine(
+                        VCFHeaderVersion.VCF4_2,
+                        line,
+                        expectedTagOrder
+                );
+            }
+            else {
+                VCFHeaderLineTranslator.parseLine(
+                        VCFHeaderVersion.VCF4_2,
+                        line,
+                        expectedTagOrder,
+                        optionalTags
+                );
+            }
+            if (!error.isEmpty()) {
+                Assert.fail("Expected failure: '" + error + "', got success");
+            }
+        }
+        catch (Exception e) {
+            if (error.isEmpty()) {
```

Comment:
Why have a value in `error` if it is unused? Is there a way to assert that the error we get is the error we expect to get?

Reply:
I am using a DataProvider for giving inputs together with an "error" field. If the error field is empty, the test was meant to succeed. If the error field is NOT empty, it tells me what kind of error I am expecting to get. That way I am simultaneously testing inputs that are okay, and those that are meant to fail. I could turn the "error" into "errorMessage" for more clarity I guess? Or do you want these sort of tests split into two parts - one for successful ones, one for failures?